### ISSUE TITLE
DG Height Must be non-negative

### ISSUE THREAD

**Issue by PreethaBS** [dotnet/wpf#5647]
.NET Core Version: 4.8
Windows version: Windows 10 Enterprise - version : 20H2
Does the bug reproduce also in WPF for .NET Framework 4.8?: Yes
Problem description:
I have an application , in which there are 10 locations and user can create people in lots to move to these locations.
the list of people/Users created in lots are displayed in datagrid.I select a user and click command button to move the user to location 1.Once he is done at location 1, I update the status in datagrid and select him to move to location 2.I repeat this process for the x number of users in lots.
On Move next - initially I update the datagrid and call await Task.Run(() => StartProcessing() to initiate task delay and based on it result (bool flag) I update the Datagrid again.This happens simultaneously, let's say different users in different locations on Movenext.

Actual behavior:
 when i have more lots and I use vertical scrollbar to select the last lot to move next, and again use vertical scroll bar to go the first lot, and during one such Update of datagrid, it throws an Exception : Height must be non-negative.
![DG-Height Must be non-negative](https://user-images.githubusercontent.com/88112134/140567972-5192be03-8068-472f-8bf3-704d9c1603f9.jpg)

 
 
 Can you please help.

**Comment by singhashish-wpf** [comment]
Can you please provide a sample app for reproduction?

**Comment by PreethaBS** [comment]
I am not able to provide a sample application. I will try to though. But the scenario is this:
1)I have an Observable collection of let's say 10 items that are loaded from table in database.The datagrid's ItemSource is binded to this collection. The DG height is set to 200.
actual dg xaml  : 
<DataGrid Grid.Row="0" Grid.Column="0" x:Name="dgLots"
                  ItemsSource="{Binding AvailableLots, Mode=TwoWay, UpdateSourceTrigger=PropertyChanged}"
                  SelectedItem="{Binding SelectedLot, Mode=TwoWay, UpdateSourceTrigger=PropertyChanged}"
                  SelectionUnit="FullRow" SelectionMode="Single" RowHeight="30"
                  IsReadOnly="True" CanUserReorderColumns="False" Margin="15,49,15,51" AutoGenerateColumns="False"
                  HorizontalScrollBarVisibility="Visible" VerticalScrollBarVisibility="Visible" >

So out of 10 , only 8 rows of items in collection are visible and for the 9th and 10th item you need to scroll to select the items.

2) In my application - I select an item(entire row) from grid and do some async processing, by calling Task.StartProcess() and on it's completion I update the status on my grid accordingly.
3) while the previous selection is processing, I can select another item from grid and initiate processing again(which follows the same procedure as above).
Basically I can select one item after another and initiate processing, and on each completion grid is updated with the latest collection information, which is read from database.
4) when  I try to select the 9th or 10th item which is not visible in view(unless scroll down), and initiate processing. In the meantime when the other selection , let's say the 2nd or 5th which has finished is updating grid..I get this height must be zero exception.
the reason when I googled was : WPF window that keeps throwing this error during a refresh
The most likely reason is that the CollectionViewSource refreshing/changing while the VirtualizingStackPanel is in the middle of Measure.

So not sure how to work around this.
The datagrid is used only for display and item selection.(No editing is allowed)

**Comment by singhashish-wpf** [comment]
duplicate of #2854

**Comment by SamBent** [comment]
@PreethaBS - can you get a scroll trace file and matching crash dump (see instructions in #2854)?

You said ``the reason when I googled was : WPF window that keeps throwing this error during a refresh
The most likely reason is that the CollectionViewSource refreshing/changing while the VirtualizingStackPanel is in the middle of Measure.``
I assume you found that quote [here](https://social.msdn.microsoft.com/Forums/office/en-US/8d7cfe9c-d9ad-4250-b760-e494e9f61bc2/height-must-be-nonnegative), but I don't think it's the reason in your case, or even in the case in the link.  

I've been looking at several reports about this.  They all have complicated or non-deterministic repros.  Yours seems much simpler, so your STF/dump information (or even better, your repro app) might really help me find the right fix.  (It's easy to fix the exception by simply replacing negative values by 0 before calling set_Height.  But it's not so clear whether that leads to other problems later on.)

It shouldn't be too hard to create a simple repro.  The exception doesn't depend on your database, but only on the pattern/timing of scrolling vs. rebuilding your AvailableLots collection.  So try a simple app that includes your DataGrid, an AvailableLots collection that contains arbitrary data ("dummy" data), and the same mechanism for processing the selected row async and updating the grid on completion (including rebuilding AvailableLots using the code in the upper-left of your screenshot).  The "processing" can probably be trivial (do nothing) - it's the way that you notify the UI thread to "update the grid" that matters.  I think.
